crates/oxc_parser/src/lexer/number.rs

```diff
@@ -55,33 +66,47 @@ fn parse_float_without_underscores(s: &str) -> Result<f64, &'static str> {
 /// these numbers are usually not long. On x84_64, FMUL has a latency of 4 clock
 /// cycles, which doesn't include addition. Some platorms support mul + add in a
 /// single instruction, but many others do not.
+///
+/// Unfortunately, this approach has the chance to overflow for excessively
+/// large numbers. In such cases, we fall back to mul_add. Note that right now
+/// we consider leading zeros as part of that length. Right now it doesn't seem
+/// worth it performance-wise to check and strip them. Further experimentation
+/// could be useful.
 #[allow(clippy::cast_precision_loss, clippy::cast_lossless)]
 fn parse_binary(s: &str) -> f64 {
-    // b'0' is 0x30 and b'1' is 0x31.
-    // So we can convert from binary digit to its value with `c & 1`.
-    // This is produces more compact assembly than `c - b'0'`.
-    // https://godbolt.org/z/1vvrK78jf
-    const fn byte_to_value(c: u8) -> u8 {
-        debug_assert!(c == b'0' || c == b'1');
-        c & 1
-    }
-    #[cfg(test)]
-    {
-        static_assertions::const_assert_eq!(byte_to_value(b'0'), 0);
-        static_assertions::const_assert_eq!(byte_to_value(b'1'), 1);
-    }
+    /// binary literals longer than this many characters have the chance to
+    /// overflow a u64, forcing us to take the slow path.
+    const MAX_FAST_BINARY_LEN: usize = 64;
 
     debug_assert!(!s.is_empty());
     debug_assert!(!s.starts_with("0b") && !s.starts_with("0B"));
 
+    if s.len() > MAX_FAST_BINARY_LEN {
+        return parse_binary_slow(s);
+    }
+
     let mut result = 0_u64;
     for c in s.as_bytes() {
         result <<= 1;
-        result |= byte_to_value(*c) as u64;
+        result |= binary_byte_to_value(*c) as u64;
     }
     result as f64
 }
 
+#[cold]
+#[inline(never)]
+fn parse_binary_slow(s: &str) -> f64 {
+    debug_assert!(!s.is_empty());
```

Comment:
Small nit: This line is unnecessary. Function will work fine for zero-length strings, and it's checked above anyway.

Reply:
Fixed